### ISSUE TITLE
[go] StringArrayOuput stack output panics when cast as StringArrayOutput from stack reference output

### ISSUE THREAD

**Issue by DavidJFelix** [pulumi/pulumi#10977]
### What happened?

I have 2 stacks and in one of them, I `ctx.Export` a `StringArrayOutput`. In the other stack, I reference the first stack and `GetOutput` and use `AsStringArrayOutput` and the pulumi program panics.

### Steps to reproduce

* Create two golang stacks
* In stack `a`, create a StringArrayOutput and export it using `ctx.Export`
* In stack `b`, `stack.GetOutput("name").AsStringArrayOutput()`

### Expected Behavior

This should be correctly cast since the types are correct.

### Actual Behavior

```
panic: interface conversion: interface {} is []interface {}, not []string
```

### Additional context

I worked around this by writing my own `ApplyT` function that works

The following code **works**

```golang
output.ApplyT(func(i interface{}) pulumi.StringArrayOutput {
  array := i.([]interface{})
  strings := make([]string, len(array))
  for i := range array {
    strings[i] = array[i].(string)
  }
  return pulumi.ToStringArray(strings).ToStringArrayOutput()
}).(pulumi.StringArrayOutput)
```

I'm not sure if casting `[]interface{}` to `[]string` is possible in go directly, which is why I'm not opening a pull request, but given that the above works and the existing code doesn't, I feel like it would be a good improvement.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by AaronFriel** [comment]
Hey @DavidJFelix, thanks for submitting this. It looks like we could probably special case the `As*ArrayOutput` methods in that file to handle `[]interface{}` for a quality of life change.

I think you've identified the heart of this issue, which is that the direct cast will panic and that is a Go language level issue:

> I'm not sure if casting []interface{} to []string is possible in go directly

The StackReference outputs will be as general as possible at each level, with the top level outputs being equivalent to `map[string]interface{}` and any nested arrays or maps will themselves be `[]interface{}` or `map[string]interface{}`. Go panics on casting `[]interface{}` to `[]string` and requires the iteration you're doing in your Apply.

I think the method you need is `AsArrayOutput` to get `[]interface{}`, however you would still need to write an Apply over it, and if you look at the definition, it already matches the first line of your Apply:

https://github.com/pulumi/pulumi/blob/adef574983fa6f4e1ab48e50ef5d1388da1efcdb/sdk/go/pulumi/types_builtins.go#L5408-L5416

As an aside, we're also actively investigating shipping generics support for the Go SDK (#9143) with no ETA yet. Our goal is to provide safe methods that cannot panic and generally improve the usability of the Go SDK.

**Comment by DavidJFelix** [comment]
Initially, I was using `AsArrayOutput` but since like you said, I already needed `ApplyT` to convert the `[]interface{}` to `[]string` I just skipped it and added the `[]interface{}` cast into my `ApplyT` function show in the original issue above.

**Comment by AaronFriel** [comment]
@DavidJFelix Yeah, I've got a draft PR to modify the helper methods (`AsStringArrayOutput()` and associated methods to do a little additional work to try to handle this case. 

You're not the first to encounter this, and I think we should err on doing a little more to prevent that panic.

**Comment by AaronFriel** [comment]
@DavidJFelix Today's release contains a fix and will allow you to use `AsStringArrayOutput` directly. I'm always thankful for folks pushing us to improving the usability of our SDKs!

**Comment by DavidJFelix** [comment]
![bitmoji](https://sdk.bitmoji.com/render/panel/5b7bff30-d7cc-4141-9f91-d77fa135057f-4a303b21-a1b5-49a2-94e1-a59c1f9ccb9c-v1.png?transparent=1&palette=1&width=246)
Tight! Love it. Happy friday.
